fixes #18476 - users have ssh keys

timogoebel opened the pull request:
This PR adds the ability for a user to have ssh keys. These keys can be used in templates and are available in ENC data.
A user without explicit permission can create/delete keys for himself.

![image](https://cloud.githubusercontent.com/assets/4107560/22982240/d370bda6-f39e-11e6-8dc1-0bc4b4824368.png)

timogoebel committed:
fixes #18476 - users have ssh keys

sean797 commented on db/migrate/20170214132230_create_ssh_keys.rb:
```diff
@@ -0,0 +1,12 @@
+  def change
+    create_table :ssh_keys do |t|
+      t.string :name, :limit => 255
+      t.string :key, :text, :limit => 5000
```
What is this text column user for?

timogoebel replied:
it stores the actual ssh key, e.g. `ssh-rsa abcdef foreman@example.com`

sean797 replied:
Isn't that stored in `key` ?

```
katello=# \d ssh_keys
                                      Table "public.ssh_keys"
   Column    |            Type             |                       Modifiers                       
-------------+-----------------------------+-------------------------------------------------------
 id          | integer                     | not null default nextval('ssh_keys_id_seq'::regclass)
 name        | character varying(255)      | 
 key         | character varying           | 
 text        | character varying           | 
 fingerprint | character varying           | 
 user_id     | integer                     | 
 created_at  | timestamp without time zone | not null
 updated_at  | timestamp without time zone | not null
Indexes:
    "ssh_keys_pkey" PRIMARY KEY, btree (id)

timogoebel replied:
Ah, good catch! Thanks.
This line should obviously read `t.text :key, :limit => 5000`

timogoebel commented:
I just realised, we already use the `SSHKey` gem as a dependency. Changing the homemade code in favor of this gem.

ares commented on app/models/user.rb:
```diff
@@ -38,6 +38,7 @@ class User < ActiveRecord::Base
   has_many :permissions,       :through => :filters
   has_many :cached_usergroup_members
   has_many :widgets, :dependent => :destroy
+  has_many :ssh_keys, :dependent => :destroy
```
It would be great if you defined a Jail class and allow this for safe mode so so we could do something like `@host.owner.ssh_keys` in provisioning and job templates. Also it would be probably handy to have a Jail definition in SshKey class as well so we could print the public key and potentially fingerprint. Another bonus points for macro that would print public keys for a given user. No push, just an idea :-)

timogoebel replied:
@ares : Good point. `owner` is currently now allowed in host's jail. I do remember, that there were some concerns about this. Can you think of some?

ares replied:
Hm, I can't think of any. The tricky thing is that sometimes the owner is user, sometimes it's user group.

timogoebel replied:
Added the Jail classes.

timogoebel commented:
Code now uses the `SSHKey` gem. Added Jail `SshKey`, `User` and `Usergroup`. `ssh_authorized_keys` now follows the law of demeter. `SshKey` now has a `length` attribute.

ohadlevy commented on app/models/ssh_key.rb:
```diff
@@ -0,0 +1,66 @@
+class SshKey < ActiveRecord::Base
+  include Authorizable
```
is there a way to reuse KeyPair model?

timogoebel replied:
I'm reluctant on this one. It might be possible, but imho we try to model something completely different. There isn't any real code duplication now and the models behave differently. If we force them into a single model, it just gets harder to maintain and test this.

ohadlevy commented on app/views/ssh_keys/_ssh_keys_tab.html.erb:
```diff
@@ -0,0 +1,45 @@
+  <% else %>
+    <tbody>
+      <tr>
+        <td class="blank-slate-pf">
```
why not use the welcome partial instead?

timogoebel replied:
As discussed on IRC, this is unfortunately not possible as the welcome partial is only shown on the index controller. The ssh key list is only rendered as a tab on the users edit page.

ohadlevy commented on webpack/assets/javascripts/foreman_ssh_keys.js:
```diff
@@ -0,0 +1,12 @@
+import $ from 'jquery';
+
+export function autofillSshKeyName() {
```
two comments:
1. missing tests for functions
2. maybe its just me, but I would love seeing it implemented via react instead ;)

timogoebel replied:
Added a test. As discussed on IRC, react is probably not worth it in this case.

bastilian commented on app/models/usergroup.rb:
```diff
@@ -112,4 +116,12 @@ def ensure_last_admin_group_is_not_deleted
   end
+
+  def ssh_keys
+    users.map(&:ssh_keys).flatten
```
You could use `users.flat_map(&:ssh_keys)` here.

dLobatog commented on app/controllers/api/v2/ssh_keys_controller.rb:
```diff
@@ -0,0 +1,51 @@
+
+      before_action :find_resource, :only => %w{show destroy}
+
+      api :GET, "/users/:user_id/ssh_keys", N_("List all ssh keys for a user")
```
nitpick, 'SSH' uppercase in user-facing strings

dLobatog commented on app/controllers/concerns/foreman/controller/ssh_keys_common.rb:
```diff
@@ -0,0 +1,25 @@
+    end
+  end
+
+  def editing_self?
```
This is duplicating logic already existing in the UsersMixin, extract to a shared concern or see if you can simply reuse the UsersMixin in both SSHKeys and User controllers

dLobatog commented on app/helpers/application_helper.rb:
```diff
@@ -127,11 +127,19 @@ def link_to_function_if_authorized(name, function, options = {}, html_options =
   end
 
   def display_delete_if_authorized(options = {}, html_options = {})
+    if authorized_for(options)
```
nitpick, a guard clause is more clear   (https://github.com/bbatsov/ruby-style-guide#no-nested-conditionals)

```ruby
return "" unless authorized_for(options)
display_delete(options, html_options)
```

dLobatog commented on app/models/host/base.rb:
```diff
@@ -57,6 +57,8 @@ def self.taxonomy_conditions
     scope :no_location,     -> { rewhere(:location_id => nil) }
     scope :no_organization, -> { rewhere(:organization_id => nil) }
 
+    delegate :ssh_authorized_keys, :to => :owner, :allow_nil => true
```
This one shouldn't delegate when the owner is an user group. Or alternatively, you can implement the method to fetch all ssh keys for all users in the user group.

```ruby
:if => ->(host) { host.owner_type == 'User' }
```

OR


```ruby
def ssh_authorized_keys
return owner.ssh_authorized_keys if owner_type == 'User'
owner.users.map(&:ssh_authorized_keys).flatten.uniq
end
```

Probably the 2nd option is better

timogoebel replied:
> Or alternatively, you can implement the method to fetch all ssh keys for all users in the user group.

I think, I did implement it for both `User` and `Usergroup`.

dLobatog commented on app/models/ssh_key.rb:
```diff
@@ -0,0 +1,66 @@
+    self.fingerprint = SSHKey.fingerprint(self.key)
+    true
+  rescue SSHKey::PublicKeyError
+    nil
```
Why?

timogoebel replied:
If I return `false` here, Rails will stop processing all further validations. The docs suggest otherwise, but the tests completely blew.

dLobatog replied:
Wouldn't it be better to display a general error if this happens as `errors.add(:base => "Could not generate fingerprint - #{e})`? 

As it stands right now if this exception happens, the user cannot know what happened, or how to fix it - not even the Administrator as it does not log anything.

timogoebel replied:
I'll add some logging. The messages provided by the gem aren't translated and are quite technical (e.g. "byte array too short"). That's why I don't want to show them to a user.

dLobatog reviewed (changes requested):
@timogoebel Sorry didn't get to review the entire thing, but I left some comments already, feel free to fix them in the meantime, I'll finish it some time next week.

IMO - if we kept the API portion in another PR it'd get easier to review 😄

dLobatog commented on app/validators/ssh_key_validator.rb:
```diff
@@ -0,0 +1,15 @@
+  def valid_ssh_public_key?(key)
+    SSHKey.valid_ssh_public_key?(key)
+  rescue SSHKey::PublicKeyError
+    false
```
I don't understand why this exception is rescued as `nil` in the model but `false` here.

timogoebel replied:
`false` is the intended and correct return value. See my other comment why I chose the other method to return `nil`.

dLobatog replied:
I see - please add a message though to the error logger when this happens - and ideally a message with the exception to display in the UI

timogoebel commented:
@bastilian, @dLobatog : Thanks for your comments. I pushed an updated version that should address them.

bastilian commented:
`.flat_map` FTW! 🎉

ohadlevy commented on app/views/ssh_keys/_form.html.erb:
```diff
@@ -0,0 +1,18 @@
+      <% end %>
+    </div>
+  <% end %>
+  <%= textarea_f f, :key, :size => "col-md-8" , :rows=> '3' %>
```
should we allow to upload ssh key as well?
also, i could see us using the ACE editor to display / edit the key.
for example, again from openshift:
![image](https://cloud.githubusercontent.com/assets/20167/23160586/bb6ff22c-f82f-11e6-8a1e-d6c365da5c4d.png)

timogoebel replied:
@ohadlevy : I think, the upload idea is good. The ACE editor doesn't bring much value imho.
Anyways, both features aren't really necessary to get the core functionality in. @dLobatog already suggested to split the API part in a separate PR. I think, we should keep this out of this PR and create follow-up tickets.

dLobatog replied:
I do agree - no need to complicate it further

dLobatog commented on app/controllers/concerns/foreman/controller/user_self_editing.rb:
```diff
@@ -0,0 +1,9 @@
+
+  protected
+
+  def editing_self?
```
Seems like this isn't very useful, it forces you to change `editing_self?`, and completely duplicate the method every time you use it (changing the sliced params).

How about:

```
def editing_self?
  @editing_self ||= User.current.editing_self?(editing_self_params)
end

def editing_self_params
  params.slice(:controller, :action, :id)
end
```

Then on SshKeysCommon,

```
def editing_self_params
  params.slice(:controller, :action).merge { :id => params[:id] } 
end
```

At that point I think you can also safely revert the changes made here to the `editing_self?` method and leave it unchanged. Any further implementation just needs to provide its own params instead of overriding the entire action

dLobatog commented on app/controllers/concerns/foreman/controller/parameters/ssh_key.rb:
```diff
@@ -0,0 +1,16 @@
+    def ssh_key_params_filter
+      Foreman::ParameterFilter.new(::SshKey).tap do |filter|
+        filter.permit :key,
+          :name
```
nitpick, I guess same line is fine for 2 keys

dLobatog commented on app/helpers/application_helper.rb:
```diff
@@ -127,11 +127,16 @@ def link_to_function_if_authorized(name, function, options = {}, html_options =
   end
 
   def display_delete_if_authorized(options = {}, html_options = {})
+    return "" unless authorized_for(options)
```
Not sure why `display_delete_if_authorized` doesn't work for you here? It seems a bit risky to modify this method to me..

timogoebel replied:
I need to display the link if either the user is authorized _or_ `@editing_self` is true.
I think, the change is quite harmless.
Before this commit, `display_delete_if_authorized` would just call `display_link_if_authorized`. The latter just calls `link_to` if `authorized_for` returns a truthy value or `""`.
Now this logic is also present in `display_delete_if_authorized`. Any better ideas are greatly appreciated, though?

timogoebel replied:
Ignore that, I found a better way to make this work. Please see `ApplicationHelper`'s `authorized_for`.

dLobatog commented on app/models/usergroup.rb:
```diff
@@ -112,4 +116,12 @@ def ensure_last_admin_group_is_not_deleted
   end
+
+  def ssh_keys
+    users.flat_map(&:ssh_keys)
```
`all_users` instead, otherwise it will not return the SSH keys for nested user groups

dLobatog commented on app/models/usergroup.rb:
```diff
@@ -112,4 +116,12 @@ def ensure_last_admin_group_is_not_deleted
+    users.flat_map(&:ssh_keys)
+  end
+
+  def ssh_authorized_keys
```
Duplicate code (the Jail and this method) between User and Usergroup, please extract to a concern

dLobatog commented on app/models/user.rb:
```diff
@@ -351,7 +356,13 @@ def editing_self?(options = {})
     options[:controller].to_s =~ /\Aapi\/v\d+\/users\Z/ &&
       options[:action] =~ /show|update/ &&
-      (options[:id].to_i == self.id || options[:id] == self.login)
+      (options[:id].to_i == self.id || options[:id] == self.login) ||
```
I think these changes are avoidable (see my previous comment on the controller)

timogoebel replied:
@dLobatog : This method looks like a code smell to me. I don't see how we can get rid of the method, though. Any suggestions?

dLobatog commented on app/models/ssh_key.rb:
```diff
@@ -0,0 +1,66 @@
+    self.length = SSHKey.ssh_public_key_bits(self.key)
+    true
+  rescue SSHKey::PublicKeyError
+    nil
```
Ditto as in https://github.com/theforeman/foreman/pull/4290/files#r102976270

dLobatog commented on app/views/ssh_keys/_ssh_keys_tab.html.erb:
```diff
@@ -0,0 +1,45 @@
+            <h1><%= _('SSH keys') %></h1>
+            <p>
+              <%= _("You can add SSH public keys to a user in Foreman.") %><br />
+              <%= _("The keys can then be used in templates and are available in config management tools managed by Foreman.") %><br />
```
nitpick, I think the 'then' here can be spared, and "config" -> "configuration", or "config.". Also "available for".

"The keys can be used in provisioning templates and are available for configuration management tools managed by Foreman."

cc @domcleal @GregSutcliffe if you can double check.. sorry

johnpmitsch replied:
@dLobatog If I'm understanding the sentence correctly, maybe add an "also" 

"The keys can be used in provisioning templates and are also available for configuration management tools managed by Foreman."

dLobatog commented on app/views/ssh_keys/_ssh_keys_tab.html.erb:
```diff
@@ -0,0 +1,45 @@
+    <% if user.ssh_keys.any? %>
+    <thead>
+      <tr>
+        <th width="40%"><%= _('Name') %></th>
```
Please use `class="col-md-x col-sm-x"` instead of fixed percentages and pixels - it's way more flexible and looks good on smaller screens

dLobatog commented on db/migrate/20170214132230_create_ssh_keys.rb:
```diff
@@ -0,0 +1,13 @@
+  def change
+    create_table :ssh_keys do |t|
+      t.string :name, :limit => 255
+      t.text :key, :limit => 5000
```
Maximum is 16384, is there any reason to limit it here?

dLobatog reviewed (changes requested):
@timogoebel Overall pretty good - I was able to test it successfully, just found a few little problems in the implementation. Good job 😄 !

timogoebel commented:
@dLobatog : Thanks, I think I addressed all your comments. Let's refactor `User.editing_self?` in a separate PR to not overcomplicate things.

timogoebel commented:
- Fixed a bug, where the focus on the SSH keys form was set to the name but not to the textarea for the key.
- The key comment is now using the user's login and the foreman url setting.
- The ENC data now contains a user's hash as well that can e.g. used in a puppet module to create users on a VM based on the owner field.

Thanks @ekohl for testing this and the great ideas.

timogoebel commented:
Rebased, renamed the `users` ENC parameter to `foreman_users`

ekohl commented:
Untested, but https://github.com/ekohl/puppet-foreman_simple_user should create the users with their SSH keys.

ekohl commented:
So yesterday we tested the module, made some changes and it works. @timogoebel also has a snippet that can be used during provisioning that implements the same logic (it's based on https://gist.github.com/ekohl/ad6863798aed19c5fb45d3b91e8a4d87 but has a fix for when the owner is a group).

@ares you might have an opinion about the templating aspect.

@GregSutcliffe I believe this is a feature you really wanted.

ares commented on app/models/usergroup.rb:
```diff
@@ -112,4 +121,8 @@ def ensure_last_admin_group_is_not_deleted
     User.unscoped.only_admin.except_hidden - all_users
   end
+
+  def ssh_keys
```
should this method be protected if it's allowed by Jail?

ares commented:
Jail definitions seems to be fine except one nit. If you point me to the template snippet, I can review that. I think it would be beneficial if it's part of community-templates too.

timogoebel commented:
@ares: Added the template to this commit. @ekohl will create a PR in community-templates as he did most of the work and deserves the credit.

- Made the protected method public.
- Fixed a typo in a Jail definition.

ares commented:
> @ares: Added the template to this commit.

Sorry if I confused you, the template is normally synced from community-templates and should not be added directly in the PR, so please remove it.

timogoebel commented:
Removed the template.
Anything more to do or is this ready to 🚢 ?

ares commented:
My comments were addressed :+1: (I didn't do the full review)

dLobatog commented:
I'll spare some time today to review this - last time it was nearly ready but I was waiting for @ares 👍

dLobatog commented:
My comments were addressed nicely too, thanks @timogoebel @ares @ekohl @ohadlevy , let's get this on track for 1.15 !
